### PR TITLE
Xiaoling/add batch gemm

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this RP, we add the following:
* the source code of the Batch Stride Gemm
* the random data random matrix size unit test of the Batch Stride Gemm
* the random data corner case unit test of the Batch Stride Gemm
* the documentation of the Batch Stride Gemm
* the golden model for Batch Stride Gemm in Scala

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -107,13 +107,71 @@ bool blockGemm(int M, int K, int N, int8_t *A, int8_t *B, int32_t * C)
 | a_i | io_data_a_i | meshRow * tileSize * input | In | A big bus containing all the data elements of the input (sub-)matrix A |
 | b_i | io_data_b_i | tileSize * meshCol * input | In | A big bus containing all the data elements of the input (sub-)matrix B |
 | c_o | io_data_c_o | meshRow * meshCol * output | Out | A big bus containing all the data elements of the input (sub-)matrix C |
+
+
+### Batch Stride GEMM
+The Batch Stride GEMM is built with the Block GEMM. It takes in an extra Batch configuration (B) except for the M, K and N. It also takes in six extra strides configuration, eg., ldA, ldB, ldC, strideA, strideB, and strideC, for computing the address for each sub-matrix in the block matrix multiplication and the start matrix for each batch.
```
Comment (JosseVanDelm) [inline README.md:113]:
```suggestion
The Batch Stride GEMM is built with the Block GEMM. Besides  M, K and N It takes in an extra Batch configuration (B). It also takes in six extra strides configuration, eg., ldA, ldB, ldC, strideA, strideB, and strideC, for computing the address for each sub-matrix in the block matrix multiplication and the start matrix for each batch. 
```

File: README.md [hunk 2]
```diff
@@ -107,13 +107,71 @@ bool blockGemm(int M, int K, int N, int8_t *A, int8_t *B, int32_t * C)
...
+
+
+### Batch Stride GEMM
+The Batch Stride GEMM is built with the Block GEMM. It takes in an extra Batch configuration (B) except for the M, K and N. It also takes in six extra strides configuration, eg., ldA, ldB, ldC, strideA, strideB, and strideC, for computing the address for each sub-matrix in the block matrix multiplication and the start matrix for each batch. 
+
+#### Computation analysis
+Batch Stride GEMM does Batch number of the same size Block Gemm.
+The computation flow is shown as the following picture.
+
+![](./docs/batch_block_matrix_mul.png)
```
Comment (JosseVanDelm) [inline README.md:119]: You did not include this file

File: src/test/scala/gemm/TestBatchGemm.scala [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+      )
+      A(k) = a
+      B(k) = b
+    }
+
+    (A, B)
+  }
+
+  // Convert the input matrix to a big bus
+  def SpliteBatchBlockMatrx(
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/TestBatchGemm.scala:61]:
```suggestion
  def SplitBatchBlockMatrx(
```

File: src/test/scala/gemm/TestBatchGemm.scala [hunk 4]
```diff
@@ -0,0 +1,526 @@
+package gemm
+
+import chisel3._
+import org.scalatest.flatspec.AnyFlatSpec
+import chiseltest._
+
+object MatrixLibBatch {
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/TestBatchGemm.scala:7]: I have the feeling we can generate a golden model class that provides a common interface for all these tests?
Right now there seems to be a lot of code duplication between all these tests?
Okay for me if we would do this in a next PR

File: src/test/scala/gemm/TestBatchGemm.scala [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+        BatchGemmRandomTest(dut, size_Batch, size_M, size_K, size_N)
+      }
+  }
+}
+
+// Simple Batch Gemm test to see if the control signals work well before random data test for debug purpose
+// In this test, the B, M, K, N are set manually to see the behavior of the Batch Gemm and to check if the control signals works well
+// Testing what if the start_do_i is asserted when the Batch Gemm is busy
+// And also orchestrating the data_valid_i to see if the Gemm works well under different situations
+// Finally checking the output manually in the waveform
```
Comment (JosseVanDelm) [inline src/test/scala/gemm/TestBatchGemm.scala:383]: Why are we testing this manually???

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]

File: src/test/scala/gemm/TestBatchGemm.scala (hunk 4 above)
Comment (xiaoling-yi) [reply src/test/scala/gemm/TestBatchGemm.scala:7]: Ok! I will try to refactor the golden model a bit.

File: src/test/scala/gemm/TestBatchGemm.scala (hunk 5 above)
Comment (xiaoling-yi) [reply src/test/scala/gemm/TestBatchGemm.scala:383]: https://www.youtube.com/watch?v=L0MK7qz13bU!

**Review by rgantonio — APPROVED** [review]
Yes! It fine but add the `ldA` stuff in the drawing or make sure it's much clear.

Think about the audience or the user if they can understand it 🌷
